### PR TITLE
feat: Phase 17 — Mermaid ER diagrams + structured FK introspection (+ Batch G roadmap)

### PR DESCRIPTION
**Pull request by devopam**
## Summary

Phase 17 of Batch A (`PLAN.md` §11). Adds the schema-visualisation
deliverable plus a piece of structured FK introspection the renderer
needed.

- **`generate_schema_diagram`** — new `mcpg.diagrams` module renders a
  schema as a Mermaid `erDiagram` block: entities with PK/FK markers
  and parent→child edges. Views and foreign tables are excluded;
  partitions are excluded by default and can be drawn with
  `include_partitions=true`. Edges pointing outside the rendered entity
  set (cross-schema or filtered-out tables) are skipped rather than
  left dangling. Output is a single string an agent can paste into any
  Mermaid-aware renderer.
- **`list_foreign_keys`** — new in `mcpg.introspection`. Gives the
  renderer structured FK data instead of regex'ing
  `pg_get_constraintdef`. The two column arrays are aligned by ordinal
  position, computed via `unnest(... WITH ORDINALITY)` joined to
  `pg_attribute` on both sides. Surfaced as a public tool — useful
  standalone for any "which columns reference what" question, and
  prerequisite for Phase 18 schema diff.

Two MCP tools registered (was 29 after Phase 16, now 31).

### Roadmap update — Batch G (ORM bridges) — also included

`PLAN.md` §11 gains **Batch G — ORM bridges** with **Phase 28
`generate_prisma_schema`** as a deliberate USP move. No other PG MCP
server bridges the catalog to an ORM's schema DSL; Prisma is the
highest-leverage starting point (~3M weekly downloads). Scope is
intentionally narrow: catalog → DSL only — no `.prisma` → DDL
parsing and no `prisma migrate` subprocess driving. Drizzle,
SQLAlchemy, and sqlc can follow as siblings under the same umbrella.

This commit only records the roadmap entry; Batch G is not in flight.

## Test plan

- [ ] CI: ruff, mypy --strict, full unit + integration suites green
- [ ] CI matrix: PG 14 / 15 / 16 / 17 — integration test renders a real
      two-table schema with a real FK and asserts the diagram output
- [ ] Coverage gate (fail_under = 90) on authored code; 100%
      maintained
- [ ] Unit tests exercise `_sanitize` edge cases (whitespace, parens,
      pathological all-punctuation), `_parse_pk_columns` (quoted +
      unquoted), partition filtering, cross-schema edge filtering, and
      MCP tool wiring via the in-process client

https://claude.ai/code/session_0122yLZLJ8t4W43sdN6BmTZc

---
_Generated by [Claude Code](https://claude.ai/code/session_0122yLZLJ8t4W43sdN6BmTZc)_

## Summary by Sourcery

Add a Mermaid-based schema diagram generator and structured foreign key introspection, and expose them as new MCP tools alongside roadmap and documentation updates.

New Features:
- Expose a list_foreign_keys introspection tool that returns structured foreign key metadata for all tables in a schema.
- Introduce a generate_schema_diagram tool that renders a Mermaid ER diagram from the database schema, with optional partition inclusion.

Enhancements:
- Add a ForeignKeyInfo data structure and query to resolve foreign key relationships with column-level detail.
- Wire the new diagram generator into the server tool registration flow so it is available in read-only access mode.

Documentation:
- Update PROGRESS.md and PLAN.md to reflect completion of Phase 17, the addition of the new tools, and the future Batch G ORM bridge roadmap.

Tests:
- Add unit tests for the diagram generator’s sanitisation, primary-key parsing, partition and cross-schema filtering, and MCP tool wiring.
- Add unit tests for list_foreign_keys to verify column alignment and MCP exposure.
- Add an integration test that renders a diagram from a real two-table schema with a foreign key edge.

### CODE REVIEW

File: src/mcpg/introspection.py [hunk 1]
```diff
@@ -131,6 +131,23 @@ class ConstraintInfo:
...
+class ForeignKeyInfo:
+    """A foreign-key constraint resolved to its referenced columns.
+
+    ``from_columns`` and ``to_columns`` are aligned by ordinal position —
+    ``from_columns[i]`` references ``to_columns[i]`` of
+    ``to_schema.to_table``.
+    """
+
+    name: str
+    from_table: str
```
[the remark this answers is not in the document]
Comment (devopam) [reply src/mcpg/introspection.py:144]: Intentionally skipping this one for codebase consistency. Every other frozen dataclass in `introspection.py` carries its plural fields as `list[str]` — `EnumInfo.values`, `RoleInfo.member_of`, `PolicyInfo.roles`, `DomainInfo.constraints`, `PublicationInfo.tables`, `SubscriptionInfo.publications`. Switching `ForeignKeyInfo` to `tuple[str, ...]` would make it the odd one out; switching all of them is a codebase-wide refactor that doesn't belong in this PR. If we decide to enforce tuple semantics for "frozen container fields", I'd rather do that in a dedicated change with the matching test updates across every type.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by devopam** [comment]
Status update — every Sourcery thread on this PR has been addressed in `4c7554c`. CI is fully green (Lint, Sourcery, Tests PG 14/15/16/17). Two threads were auto-recognized; the remaining six contain the fix but Sourcery didn't auto-resolve them because the new test names diverge from its suggested literals:

| Comment | Addressed by |
| --- | --- |
| `tests/unit/test_diagrams.py:25` — whitespace `_sanitize` case | `test_sanitize_collapses_non_alphanumeric_runs_and_strips_edges` now asserts `_sanitize("  text with spaces  ") == "text_with_spaces"` |
| `tests/unit/test_diagrams.py:100` — positive `include_partitions=True` test | new `test_generate_schema_diagram_includes_partitions_when_requested` |
| `tests/unit/test_diagrams.py:123` — filtered-partition edge skip | new `test_generate_schema_diagram_skips_edges_pointing_to_filtered_partitions` |
| `tests/unit/test_diagrams.py:179` — MCP wiring assertion shape | `test_generate_schema_diagram_tool_is_registered_and_callable` now asserts `result.content[0].text.startswith("erDiagram\n")` |
| `tests/integration/test_diagrams_integration.py` — tighter FK assertion | `assert "integer widget_id FK" in rendered` |
| `tests/unit/test_introspection.py:270` — empty `list_foreign_keys` test | new `test_list_foreign_keys_returns_an_empty_list_when_no_foreign_keys` |

PR is ready for merge.

---
_Generated by [Claude Code](https://claude.ai/code)_
